modules/programs/neovim.nix

```diff
@@ -394,26 +393,32 @@ in {
 
     home.sessionVariables = mkIf cfg.defaultEditor { EDITOR = "nvim"; };
 
-    xdg.configFile =
-      let hasLuaConfig = hasAttr "lua" config.programs.neovim.generatedConfigs;
-      in mkMerge (
-        # writes runtime
-        (map (x: x.runtime) pluginsNormalized) ++ [{
-          "nvim/init.lua" = let
-            luaRcContent =
-              lib.optionalString (neovimConfig.neovimRcContent != "")
-              "vim.cmd [[source ${
+    xdg.configFile = mkMerge (
+      # writes runtime
+      (map (x: x.runtime) pluginsNormalized) ++ [{
+        "nvim/init.lua" = let
+          vimlConfig =
+            lib.optionalString (neovimConfig.neovimRcContent != "") ''
+              vim.cmd [[source ${
                 pkgs.writeText "nvim-init-home-manager.vim"
                 neovimConfig.neovimRcContent
-              }]]" + config.programs.neovim.extraLuaConfig
-              + lib.optionalString hasLuaConfig
-              config.programs.neovim.generatedConfigs.lua;
-          in mkIf (luaRcContent != "") { text = luaRcContent; };
-
-          "nvim/coc-settings.json" = mkIf cfg.coc.enable {
-            source = jsonFormat.generate "coc-settings.json" cfg.coc.settings;
-          };
-        }]);
+              }]]
+            '';
+          luaPluginsConfig =
+            lib.optionalString (hasAttr "lua" cfg.generatedConfigs)
+            cfg.generatedConfigs.lua;
+          luaRcContent = builtins.concatStringsSep "\n"
+            (builtins.filter (str: str != "") [
+              vimlConfig
+              cfg.extraLuaConfig
+              luaPluginsConfig
+            ]);
```

Comment:
If we're going to do this, maybe we can even add headers as comments above each section?

```nix
''
  -- Legacy vim configuration
  ${optionalString ...}

  -- Lua configuration
  ...
''
```

Reply:
Usually I don't like comments with empty sections or more than two consecutive newlines (which is often a result from optional strings written this way). If I skip comments for empty strings, then it would be strange to have a comment with only one section, no? Also it is very important to not create any configs if all sections are empty.

It you think that this is a right way, I can try to do it.